core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt

```diff
@@ -0,0 +1,1218 @@
+    ) : MyMarker(
+        name = name,
+        visibility = visibility,
+        fields = fields,
+        superMarkers = emptyList(),
+    ) {
+
+        // enums become List<Something>, not Dataframe<*>
+        override val isPrimitive: Boolean = true
+
+        override fun withName(name: String): Enum =
+            Enum(nullable, fields, name, visibility)
+
+        override fun withVisibility(visibility: MarkerVisibility): Enum =
+            Enum(nullable, fields, name, visibility)
+    }
+
+    class Interface(
+        fields: List<GeneratedField>,
+        superMarkers: List<Marker>,
+        name: String,
+        visibility: MarkerVisibility = MarkerVisibility.IMPLICIT_PUBLIC,
+    ) : MyMarker(
+        name = name,
+        visibility = visibility,
+        fields = fields,
+        superMarkers = superMarkers,
+    ) {
+
+        override val isPrimitive = false
+
+        override fun withName(name: String): Interface =
+            Interface(fields, superMarkers.values.toList(), name, visibility)
+
+        override fun withVisibility(visibility: MarkerVisibility): Interface =
+            Interface(fields, superMarkers.values.toList(), name, visibility)
+    }
+
+    /** Type alias that points at something other than a Marker. */
+    class TypeAlias(
+        val superMarker: Marker,
+        name: String,
+        visibility: MarkerVisibility = MarkerVisibility.IMPLICIT_PUBLIC,
+    ) : MyMarker(
+        name = name,
+        visibility = visibility,
+        fields = emptyList(),
+        superMarkers = listOf(superMarker),
+    ) {
+
+        override val isPrimitive = true
+
+        override fun withName(name: String): TypeAlias = TypeAlias(superMarker, name, visibility)
+
+        override fun withVisibility(visibility: MarkerVisibility): TypeAlias =
+            TypeAlias(superMarker, name, visibility)
+    }
+
+    /** Type alias that points at another Marker. */
+    class MarkerAlias(
+        val superMarker: Marker,
+        name: String,
+        visibility: MarkerVisibility = MarkerVisibility.IMPLICIT_PUBLIC,
+    ) : MyMarker(
+        name = name,
+        visibility = visibility,
+        fields = emptyList(),
+        superMarkers = listOf(superMarker),
+    ) {
+
+        override val isPrimitive = false
+
+        override fun withName(name: String): MarkerAlias = MarkerAlias(superMarker, name, visibility)
+
+        override fun withVisibility(visibility: MarkerVisibility): MarkerAlias =
+            MarkerAlias(superMarker, name, visibility)
+    }
+}
+
+/**
+ * Converts named OpenApi schemas to a list of [MyMarker]s.
+ * Will cause an exception for circular references, however they shouldn't occur in OpenApi specs.
+ */
+private fun Map<String, Schema<*>>.toMarkers(): List<MyMarker> {
+    // Convert the schemas to toMarker calls that can be repeated to resolve references.
+    val retrievableMarkers = mapValues { (typeName, value) ->
```

Comment:
Can you explain this part with references, resolve, retrieve and  ProduceAdditionalMarker? It seems i don't know enough about open api to understand it

Reply:
This is actually something I came up with myself. OpenApi provides schemas for all of the types used. For each type, we want to generate a Marker (Which can be an interface, enum, etc.). However, the OpenApi schema is not ordered per se, so when we are reading the schema it might be that we have a reference to a (super)type (which are queried using `getRefMarker`) for which we have not yet created a Marker. In that case, I "pause" that one (by returning CannotFindRefMarker) and try to read another type schema first. Circular references cannot exist since it's encoded in JSON, so we never get stuck in an infinite loop. When all markers are "retrieved" (so turned from a RetrievableMarker to a MarkerResult.Success), we're done and have converted everything!

As for `produceAdditionalMarker`: In OpenAPI not all enums/objects have to be defined as a separate schema. Although recommended, you can still define an object anonymously directly as a type. For this, I have `produceAdditionalMarker` since during the conversion of a schema->Marker we get an additional new Marker.